File: compilers/zig/src/tests/n095_synthetic_array_chain.zig
//! N-095 — the synthetic-array chain on `ANFProperty` is wire data.
//!
//! The expand-fixed-arrays pass desugars a `FixedArray` property into scalar
//! siblings and hangs a chain of `{base, index, length}` levels off each leaf.
//! `regroupStateFields` in `codegen/emit.zig` collapses those siblings back
//! into a single FixedArray state entry by reading that chain off the ANF
//! PROGRAM, not off the AST. So the field is load-bearing on the wire, and it
//! is load-bearing in a way nothing in this repo was watching: drop it and the
//! script hex is byte-identical, only the artifact's `stateFields` degrades
//! from one `grid` entry into four raw `grid__i__j` scalars, which is what the
//! SDK's `state.grid` accessor is built on.
//!
//! Zig's defect was the quietest of the three shapes:
//!   * it threaded the chain correctly through the AST, `expand_fixed_arrays`
//!     and `anf_lower`, so SOURCE mode regrouped fine; but
//!   * `ir/json.zig` neither wrote it in `writePropertiesArray` nor read it in
//!     `parseProperties`, so `--emit-ir` dropped it on the floor and
//!     `compile-ir` could not recover it from anyone's ANF, its own included.
//!
//! Three tiers spelled the key three different ways (`syntheticArrayChain`,
//! `__syntheticArrayChain`, `synthetic_array_chain`) and three emitted nothing.
//! The settled spelling is Go's `syntheticArrayChain`, which is also what
//! `$defs.ANFProperty` in the shared JSON Schema now declares.

const std = @import("std");
const compiler_api = @import("../compiler_api.zig");
const json_ir = @import("../ir/json.zig");
const types = @import("../ir/types.zig");
const emit = @import("../codegen/emit.zig");
const stack_lower = @import("../passes/stack_lower.zig");
const peephole = @import("../passes/peephole.zig");

/// Nested 2x2 FixedArray: four leaves, each with a two-level chain.
const GRID_SRC =
    \\import { StatefulSmartContract, assert } from 'runar-lang';
    \\class Grid2x2 extends StatefulSmartContract {
    \\  grid: FixedArray<FixedArray<bigint, 2>, 2> = [[0n, 0n], [0n, 0n]];
    \\  constructor() {
    \\    super();
    \\  }
    \\  public set00(v: bigint) {
    \\    this.grid[0][0] = v;
    \\    assert(true);
    \\  }
    \\  public set11(v: bigint) {
    \\    this.grid[1][1] = v;
    \\    assert(true);
    \\  }
    \\}
;

/// Control: no FixedArray anywhere, so no property may grow the key.
const SCALAR_SRC =
    \\import { StatefulSmartContract, assert } from 'runar-lang';
    \\class Counter extends StatefulSmartContract {
    \\  count: bigint = 0n;
    \\  constructor() {
    \\    super();
    \\  }
    \\  public increment() {
    \\    this.count = this.count + 1n;
    \\    assert(true);
    \\  }
    \\}
;

/// Run passes 1-4 and hand back the canonical ANF JSON `--emit-ir` prints.
/// Caller owns the returned bytes.
/// `work` must be an ARENA: `runPipeline` allocates every node from it and
/// the ANF types carry no recursive `deinit`, so the arena is the only
/// correct owner. Every test below follows `ir/json.zig`'s own convention.
fn emitIr(work: std.mem.Allocator, source: []const u8) ![]const u8 {
    var diag = compiler_api.Diagnostics{};
    const pipeline = try compiler_api.runPipeline(work, source, "Contract.runar.ts", .{
        .stop_after = .anf,
        .disable_constant_folding = true,
    }, &diag);
    const program = pipeline.program orelse return error.NoProgram;
    return json_ir.serializeCanonicalJSON(work, program);
}

/// Compile a parsed ANF program the way `compile-ir` does, returning the
/// artifact JSON. Caller owns the returned bytes.
fn artifactFromProgram(allocator: std.mem.Allocator, program: types.ANFProgram) ![]const u8 {
    const stack_program = try stack_lower.lower(allocator, program);
    const optimized = try peephole.optimize(allocator, stack_program.methods);
    return emit.emitArtifact(allocator, .{
        .methods = optimized,
        .contract_name = stack_program.contract_name,
        .properties = stack_program.properties,
        .constructor_params = stack_program.constructor_params,
    }, program);
}

// ---------------------------------------------------------------------------
// Wire format
// ---------------------------------------------------------------------------

test "N-095: --emit-ir writes syntheticArrayChain on every expanded leaf" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const ir = try emitIr(allocator, GRID_SRC);

    // The settled spelling, and only it.
    try std.testing.expect(std.mem.indexOf(u8, ir, "\"syntheticArrayChain\"") != null);
    try std.testing.expect(std.mem.indexOf(u8, ir, "\"__syntheticArrayChain\"") == null);
    try std.testing.expect(std.mem.indexOf(u8, ir, "\"synthetic_array_chain\"") == null);

    const parsed = try std.json.parseFromSlice(std.json.Value, allocator, ir, .{});
    defer parsed.deinit();
    const props = parsed.value.object.get("properties").?.array;
    try std.testing.expectEqual(@as(usize, 4), props.items.len);

    const want = [_][2]i64{ .{ 0, 0 }, .{ 0, 1 }, .{ 1, 0 }, .{ 1, 1 } };
    for (props.items, 0..) |prop, i| {
        const chain = prop.object.get("syntheticArrayChain").?.array;
        try std.testing.expectEqual(@as(usize, 2), chain.items.len);
        try std.testing.expectEqualStrings("grid", chain.items[0].object.get("base").?.string);
        try std.testing.expectEqual(want[i][0], chain.items[0].object.get("index").?.integer);
        try std.testing.expectEqual(@as(i64, 2), chain.items[0].object.get("length").?.integer);
        try std.testing.expectEqual(want[i][1], chain.items[1].object.get("index").?.integer);
        try std.testing.expectEqual(@as(i64, 2), chain.items[1].object.get("length").?.integer);
    }
}

test "N-095: every emitted ANFProperty key is one the shared schema declares" {
    // Hand-mirrored from `$defs.ANFProperty` in
    // packages/runar-ir-schema/src/schemas/anf-ir.schema.json, which is
    // `additionalProperties: false` — any key outside this set makes Zig's own
    // ANF fail `validateANF`.
    const allowed = [_][]const u8{ "name", "type", "readonly", "initialValue", "syntheticArrayChain" };

    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const ir = try emitIr(allocator, GRID_SRC);

    const parsed = try std.json.parseFromSlice(std.json.Value, allocator, ir, .{});
    defer parsed.deinit();
    for (parsed.value.object.get("properties").?.array.items) |prop| {
        var it = prop.object.iterator();
        while (it.next()) |entry| {
            var ok = false;
            for (allowed) |a| {
                if (std.mem.eql(u8, a, entry.key_ptr.*)) ok = true;
            }
            try std.testing.expect(ok);
        }
    }
}

test "N-095: a FixedArray-free contract's ANF does not grow the key" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const ir = try emitIr(allocator, SCALAR_SRC);
    try std.testing.expect(std.mem.indexOf(u8, ir, "ynthetic") == null);
}

// ---------------------------------------------------------------------------
// The harm: the ABI, not the hex
// ---------------------------------------------------------------------------

test "N-095: Zig's own ANF replayed through compile-ir still regroups" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const ir = try emitIr(allocator, GRID_SRC);
    const program = try json_ir.parseANFProgram(allocator, ir);

    // The loader must recover the chain, not silently drop it.
    try std.testing.expectEqual(@as(usize, 4), program.properties.len);
    for (program.properties) |p| {
        const chain = p.synthetic_array_chain orelse return error.ChainLostOnLoad;
        try std.testing.expectEqual(@as(usize, 2), chain.len);
        try std.testing.expectEqualStrings("grid", chain[0].base);
    }

    const artifact = try artifactFromProgram(allocator, program);

    // One regrouped `grid`, not four raw scalars.
    try std.testing.expect(std.mem.indexOf(u8, artifact, "\"stateFields\":[{\"name\":\"grid\"") != null);
    try std.testing.expect(std.mem.indexOf(u8, artifact, "\"syntheticNames\":[\"grid__0__0\",\"grid__0__1\",\"grid__1__0\",\"grid__1__1\"]") != null);
}

test "N-095: the chain is ABI-only — the round trip must not move a script byte" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const direct = try compiler_api.compileSourceWithOptions(allocator, GRID_SRC, "Grid2x2.runar.ts", true);

    const ir = try emitIr(allocator, GRID_SRC);
    const program = try json_ir.parseANFProgram(allocator, ir);
    const artifact = try artifactFromProgram(allocator, program);

    const round_hex = try compiler_api.extractArtifactScript(artifact);
    try std.testing.expectEqualStrings(direct.script_hex, round_hex);
}

File: compilers/rust/tests/n095_synthetic_array_chain.rs
//! N-095: the synthetic-array chain on `ANFProperty` is wire data, and the
//! wire is governed by `packages/runar-ir-schema/src/schemas/anf-ir.schema.json`.
//!
//! Rust used to serialize it as `__syntheticArrayChain`. The `__` prefix is a
//! TypeScript *AST* convention for compiler-internal annotations
//! (`PropertyNode.__syntheticArrayChain`); leaking it onto the ANF JSON was an
//! AST marker escaping its layer. Three tiers spelled the field three
//! different ways and `$defs.ANFProperty` is `additionalProperties: false`, so
//! every tier that emitted it failed its own schema, and no tier could read
//! another's. The settled spelling is Go's `syntheticArrayChain`.
//!
//! The field is load-bearing: the artifact assembler regroups the expanded
//! scalar leaves back into a single FixedArray state/ABI entry by reading the
//! chain off the ANF program, not off the AST. An ANF that loses it compiles
//! to byte-identical script but degrades the SDK's `state.grid` accessor into
//! four raw scalars.

use runar_compiler_rust::{compile_from_ir_str, compile_from_source_str, compile_source_str_to_ir};
use serde_json::Value;

const GRID_SOURCE: &str = r#"import { StatefulSmartContract, assert } from 'runar-lang';
import type { FixedArray } from 'runar-lang';

export class Grid2x2 extends StatefulSmartContract {
  grid: FixedArray<FixedArray<bigint, 2>, 2> = [[0n, 0n], [0n, 0n]];

  constructor() {
    super();
  }

  public set00(v: bigint) {
    this.grid[0][0] = v;
    assert(true);
  }

  public set11(v: bigint) {
    this.grid[1][1] = v;
    assert(true);
  }
}
"#;

const SCALAR_SOURCE: &str = r#"import { StatefulSmartContract, assert } from 'runar-lang';

export class Counter extends StatefulSmartContract {
  count: bigint = 0n;

  constructor() {
    super();
  }

  public increment() {
    this.count = this.count + 1n;
    assert(true);
  }
}
"#;

fn emit_ir(source: &str, file_name: &str) -> Value {
    let program = compile_source_str_to_ir(source, Some(file_name)).expect("compile to IR");
    serde_json::to_value(&program).expect("serialize ANF")
}

/// Every key Rust writes on an ANFProperty must be one the cross-tier schema
/// declares. `$defs.ANFProperty` is `additionalProperties: false`, so an
/// undeclared key makes Rust's own ANF fail `validateANF`.
#[test]
fn emitted_property_keys_are_all_declared_in_the_schema() {
    let schema_path = concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../../packages/runar-ir-schema/src/schemas/anf-ir.schema.json"
    );
    let schema: Value =
        serde_json::from_str(&std::fs::read_to_string(schema_path).expect("read ANF schema"))
            .expect("parse ANF schema");
    let def = &schema["$defs"]["ANFProperty"];
    assert_eq!(
        def["additionalProperties"],
        Value::Bool(false),
        "$defs.ANFProperty is no longer additionalProperties:false — this test's \
         premise (an undeclared key is a schema violation) no longer holds"
    );
    let allowed: Vec<&String> = def["properties"]
        .as_object()
        .expect("$defs.ANFProperty.properties")
        .keys()
        .collect();

    let ir = emit_ir(GRID_SOURCE, "Grid2x2.runar.ts");
    for prop in ir["properties"].as_array().expect("properties") {
        for key in prop.as_object().expect("property object").keys() {
            assert!(
                allowed.contains(&key),
                "emitted ANFProperty key {:?} is not declared in $defs.ANFProperty; \
                 allowed = {:?}",
                key,
                allowed
            );
        }
    }
}

/// Pins the spelling and the shape of every level.
#[test]
fn expanded_leaves_carry_the_camel_case_chain() {
    let ir = emit_ir(GRID_SOURCE, "Grid2x2.runar.ts");
    let props = ir["properties"].as_array().expect("properties");
    assert_eq!(props.len(), 4, "expected 4 expanded leaves");

    let want = [(0, 0), (0, 1), (1, 0), (1, 1)];
    for (i, prop) in props.iter().enumerate() {
        let obj = prop.as_object().expect("property object");
        assert!(
            !obj.contains_key("__syntheticArrayChain"),
            "leaf {i} still carries the AST-marker spelling `__syntheticArrayChain`"
        );
        assert!(
            !obj.contains_key("synthetic_array_chain"),
            "leaf {i} still carries the snake spelling `synthetic_array_chain`"
        );
        let chain = obj
            .get("syntheticArrayChain")
            .unwrap_or_else(|| panic!("leaf {i} has no `syntheticArrayChain` key: {obj:?}"))
            .as_array()
            .expect("chain is an array");
        assert_eq!(chain.len(), 2, "leaf {i}: a 2x2 grid nests twice");
        assert_eq!(chain[0]["base"], "grid", "leaf {i} outer base");
        assert_eq!(chain[0]["index"], want[i].0, "leaf {i} outer index");
        assert_eq!(chain[0]["length"], 2, "leaf {i} outer length");
        assert_eq!(chain[1]["index"], want[i].1, "leaf {i} inner index");
        assert_eq!(chain[1]["length"], 2, "leaf {i} inner length");
    }
}

/// Byte-neutrality control: a contract with no FixedArray must not grow the key.
#[test]
fn scalar_property_carries_no_chain() {
    let ir = emit_ir(SCALAR_SOURCE, "Counter.runar.ts");
    for prop in ir["properties"].as_array().expect("properties") {
        let obj = prop.as_object().expect("property object");
        assert!(
            !obj.keys().any(|k| k.to_lowercase().contains("synthetic")),
            "a FixedArray-free contract grew a synthetic-array key: {obj:?}"
        );
    }
}

/// The acceptance test: Rust's own ANF, fed back through `--ir`, must still
/// produce the regrouped `grid` state field rather than four raw scalars.
#[test]
fn self_ir_round_trip_still_regroups() {
    let ir = emit_ir(GRID_SOURCE, "Grid2x2.runar.ts");
    let ir_json = serde_json::to_string(&ir).expect("re-serialize ANF");

    let from_source = compile_from_source_str(GRID_SOURCE, Some("Grid2x2.runar.ts"))
        .expect("compile from source");
    let from_ir = compile_from_ir_str(&ir_json).expect("compile from IR");

    for (label, artifact) in [("source", &from_source), ("ir", &from_ir)] {
        assert_eq!(
            artifact.state_fields.len(),
            1,
            "{label}: expected 1 regrouped state field, got {:?}",
            artifact
                .state_fields
                .iter()
                .map(|f| &f.name)
                .collect::<Vec<_>>()
        );
        let sf = &artifact.state_fields[0];
        assert_eq!(sf.name, "grid", "{label}: state field name");
        let fa = sf
            .fixed_array
            .as_ref()
            .unwrap_or_else(|| panic!("{label}: state field carries no fixedArray metadata"));
        assert_eq!(
            fa.synthetic_names,
            vec!["grid__0__0", "grid__0__1", "grid__1__0", "grid__1__1"],
            "{label}: syntheticNames"
        );
    }

    assert_eq!(
        from_source.script, from_ir.script,
        "script diverged between source mode and IR mode"
    );
}

/// Cross-tier direction that mattered most in practice: an ANF produced by any
/// other tier under the settled spelling must regroup here too. Go's bytes are
/// the reference, so replay its exact shape.
#[test]
fn foreign_anf_with_the_settled_spelling_regroups() {
    let mut ir = emit_ir(GRID_SOURCE, "Grid2x2.runar.ts");
    // Re-key each chain through a plain JSON rebuild, i.e. exactly what a
    // foreign tier hands us: no Rust types involved.
    let props = ir["properties"].as_array().expect("properties").clone();
    let rebuilt: Vec<Value> = props
        .iter()
        .map(|p| {
            let obj = p.as_object().expect("property object");
            let mut out = serde_json::Map::new();
            out.insert("name".into(), obj["name"].clone());
            out.insert("type".into(), obj["type"].clone());
            out.insert("readonly".into(), obj["readonly"].clone());
            if let Some(v) = obj.get("initialValue") {
                out.insert("initialValue".into(), v.clone());
            }
            out.insert(
                "syntheticArrayChain".into(),
                obj["syntheticArrayChain"].clone(),
            );
            Value::Object(out)
        })
        .collect();
    ir["properties"] = Value::Array(rebuilt);

    let artifact = compile_from_ir_str(&serde_json::to_string(&ir).expect("serialize"))
        .expect("compile from foreign IR");
    assert_eq!(artifact.state_fields.len(), 1);
    assert_eq!(artifact.state_fields[0].name, "grid");
    assert!(artifact.state_fields[0].fixed_array.is_some());
}
